Ensure SimulationRunnerIterator saves states when tearing down

fjclark opened the pull request:
Modified the SimulationRunnerIterator to ensure that the state of sub-simulation runners is saved before they are torn down. This means that setting the equilibration time at the CalcSet level will now work as expected (previously the state of the calculations was not saved, causing the analysis to fail). Also ensured that methods defined at the SimulationRunner level which are inappropriately called at the CalcSet level raise NotImplemented errors.

This closes (hopefully) https://github.com/michellab/a3fe/issues/21.

## Status
- [x] Ready to go

Could you please take a quick look @Roy-Haolin-Du?

fjclark commented:
Another example of why we need those pre-commit hooks... I'll do this on the `feature-decouple-somd-2` branch soon.

Roy-Haolin-Du commented on a3fe/run/calc_set.py:
from pandas.core.api import DataFrame as _DataFrame
or
from pandas.core.api import DataFrame

fjclark commented:
Thanks for the review @Roy-Haolin-Du! Thanks for pointing out that error - turns out that that line was redundant anyway, so I've removed it. I've also added a tiny bit of code to address the issue mentioned at the end of https://github.com/michellab/a3fe/issues/21. No need to review as it's very minor - I'll merge in once the CI runs.
